### PR TITLE
Have throttled worker always wait throttleDelay between work calls

### PR DESCRIPTION
**Pull request by lramos15**
Consider the following case, `maxWorkChunkSize` of 1 and a `throttleDelay` of 1000ms. The `maxBufferedWork` doesn't matter here.

1. Queue a unit at 0ms. It immediately executes as there is nothing in the throttler
2. Queue a unit at 500ms. It immediately executes again as there is nothing in the throttler.

This seems wrong as I don't want my work items to execute without a minimum 1000ms gap in between them. Instead even if the queue is empty, the throttled worker should ensure that the second queue task must wait 500ms before executing so it's been 1000ms since the last unit of work.

This does modify the text behavior slightly, so I am happy to make this an option on the worker options or something, but I was surprised when the worker did not work out of the box like this

### CODE REVIEW

**Review by kalebers — APPROVED** [review]

File: src/vs/base/common/async.ts [hunk 1]
```diff
@@ -6,12 +6,12 @@
 import { CancellationToken, CancellationTokenSource } from './cancellation.js';
 import { BugIndicatingError, CancellationError } from './errors.js';
 import { Emitter, Event } from './event.js';
+import { Lazy } from './lazy.js';
```
Comment (bpasero) [inline src/vs/base/common/async.ts:9]: @lramos15 please turn of sorting of imports

Comment (bpasero) [inline src/vs/base/common/async.ts:9]: Sorting of imports seems irrelevant to this task. Not everyone in the team sorts imports and generally I think we should either sort all imports in the project and have a rule for it or not sort at all.

File: src/vs/base/common/async.ts [hunk 2]
```diff
@@ -1225,11 +1226,19 @@ export class ThrottledWorker<T> extends Disposable {
...
-		if (!this.throttler.value) {
+		if (!this.throttler.value && timeSinceLastExecution >= this.options.throttleDelay) {
 			this.doWork();
+		} else {
+			this.throttler.value = new RunOnceScheduler(() => {
+				this.throttler.clear();
+				this.doWork();
+			}, Math.max(this.options.throttleDelay - timeSinceLastExecution, 0));
+			this.throttler.value.schedule();
 		}
```
Comment (bpasero) [inline src/vs/base/common/async.ts:1242]: I think the condition here is dangerous where we may end up constantly scheduling a `RunOnceScheduler` if a spammy worker calls `work` with a timeout of `0` always. In other words, if someone manages to call `work` all the time before the scheduler was able to execute, the scheduler gets always re-created and never scheduled.

**Review by bpasero — CHANGES_REQUESTED** [review]

**Comment by bpasero** [comment]
I think this change would not be suitable for the case I created the `ThrottledWorker` for: if now a file watcher sends 1 event every `10ms`, we would artificially wait `200ms` before letting this event go. The intent was to start throttling when an unusual large number of events comes in at the same time:

https://github.com/microsoft/vscode/blob/b591f33513575b89cfe3f42d5b9de43efcf9ca78/src/vs/platform/files/node/watcher/parcel/parcelWatcher.ts#L176-L183

**Comment by lramos15** [comment]
Would it be worth a throttled worker option or do you think I should just use a different utility? 

Here's the PR where I try to consume this if interested. https://github.com/microsoft/vscode-copilot/pull/8758

File: src/vs/base/common/async.ts (hunk 1 above)
Comment (lramos15) [reply src/vs/base/common/async.ts:9]: Ugh I think this was caused by format on save.

File: src/vs/base/common/async.ts (hunk 2 above)
Comment (lramos15) [reply src/vs/base/common/async.ts]: This is a good point, I've made a check to make sure that if there is already a scheduled run in the throttler we don't schedule another one. Instead it will just be picked up by the next doWork call

**Review by bpasero — DISMISSED** [review]
@lramos15 yeah we can have it with that option, pushed some cleanup.

Didn't you have a test for this in your first PR?

**Comment by lramos15** [comment]
I did not create a test for it in my first PR because it modified the default behavior so instead I modified your existing PR. I will create a test for this and get another ✔️ since you did the cleanup. Thanks for the 💄 and review

**Review by joyceerhl — APPROVED** [review]
